pkg/k8s/client/restConfig_provider.go

```diff
@@ -215,3 +262,137 @@ func (r *restConfigManager) WrapRoundTripper(rt http.RoundTripper) http.RoundTri
 	r.rt.delegate = rt
 	return r.rt
 }
+
+func (r *restConfigManager) startK8sAPIServerFileWatcher() error {
+	if finfo, err := os.Stat(K8sAPIServerFilePath); errors.Is(err, os.ErrNotExist) {
+		_, err := os.Create(K8sAPIServerFilePath)
+		if err != nil {
+			return fmt.Errorf("unable to create '%s': %w", K8sAPIServerFilePath, err)
+		}
+	} else if finfo.Size() != 0 {
+		// Restore the service and endpoint addresses.
+		r.updateK8sAPIServerURL()
+	}
+
+	watcher, err := fswatcher.New([]string{K8sAPIServerFilePath})
+	if err != nil {
+		return fmt.Errorf("unable to add file watcher '%s': %w", K8sAPIServerFilePath, err)
+	}
+
+	r.jobs.Add(job.OneShot("kube-apiserver-state-file-watcher", func(ctx context.Context, health cell.Health) error {
+		health.OK("Starting kube-apiserver state file watcher")
+		return r.k8sAPIServerFileWatcher(ctx, watcher, health)
+	}))
+
+	return nil
+}
+
+func (r *restConfigManager) k8sAPIServerFileWatcher(ctx context.Context, watcher *fswatcher.Watcher, health cell.Health) error {
+	for {
+		select {
+		case <-ctx.Done():
+			health.Stopped("Context done")
+			watcher.Close()
+			return nil
+		case event := <-watcher.Events:
+			if !event.Op.Has(fsnotify.Write) {
+				continue
+			}
+			r.log.WithField("file", K8sAPIServerFilePath).Info("Processing write event ")
+			r.updateK8sAPIServerURL()
+		case err := <-watcher.Errors:
+			health.Degraded(fmt.Sprintf("Failed to load  %q", K8sAPIServerFilePath), err)
+			r.log.WithError(err).Error("Unexpected error while watching kube-apiserver state file")
+			return err
```

Comment:
So this means that if we get an error from the FS watcher, updates to the k8s API server are broken until the agent restarts.

A few things about this. First, the watcher can send and error, and still continue working. Only when the `watcher.Events` or `watcher.Errors` channel closes, has it actually stopped.

Second, might I suggest moving the creation of the FS watcher to inside of this job. And attempting to recreate the watcher if we error. With a timeout of course. That way, we have a chance to automatically recover if, for example, the file got deleted and then later re-create. 

Third, even if you want to keep this behavior, then you should still call `watcher.Close()`.

Reply:
>might I suggest moving the creation of the FS watcher to inside of this job. And attempting to recreate the watcher if we error. 

That's a good suggestion. Done!

>First, the watcher can send and error, and still continue working. Only when the watcher.Events or watcher.Errors channel closes, has it actually stopped.

Gotcha! Updated the logic to not close the watcher on errors.